workload/schemachanger: enable inserts inside workload

fqazi opened the pull request:
Previously, the schema changer workload did not have inserts
enabled, which meant our coverage was limited to operating
schema changes only on empty tables. This was inadequate
because it limited the types of issues we can detect, since
for example backfills would not be very interesting. To address,
this patch enables the insert portion of the workload and as
a part of this work it adds support for dealing with constraints,
generated columns/indexes properly.

Release justification: low risk and a workload enhancement.
Release note: None

fqazi committed:
workload/schemachanger: enable inserts inside workload

Previously, the schema changer workload did not have inserts
enabled, which meant our coverage was limited to operating
schema changes only on empty tables. This was inadequate
because it limited the types of issues we can detect, since
for example backfills would not be very interesting. To address,
this patch enables the insert portion of the workload and as
a part of this work it adds support for dealing with constraints,
generated columns/indexes properly.

Release justification: low risk and a workload enhancement.
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/78259)
<!-- Reviewable:end -->

ajwerner commented:
Seems like this is flakey under stress

fqazi commented:
> Seems like this is flakey under stress

Seems to be a build environment issue:  ERROR: geos: no locations to init GEOS causing failures :(

postamar commented:
This is tracked by https://github.com/cockroachdb/cockroach/issues/63463

ajwerner commented on pkg/workload/schemachange/error_screening.go:
```diff
@@ -205,76 +209,225 @@ SELECT array_agg(column_name)
+				if !errors.As(err, &pgErr) {
+					return false, nil, err
+				}
+				// Only accept know error types for generated expressions.
```
`s/know/known/`?

ajwerner commented on pkg/workload/schemachange/error_screening.go:
```diff
@@ -328,74 +481,280 @@ SELECT count(*) > 0
 }
 
-func (og *operationGenerator) violatesUniqueConstraintsHelper(
+// isValidGenerationError these codes can be observed when genearting an expressions.
```
genearting

Maybe say a tad more about why these might happen?

ajwerner commented on pkg/workload/schemachange/error_screening.go:
```diff
@@ -328,74 +481,280 @@ SELECT count(*) > 0
+		}
+		if isNull && !isNullable && !nullViolationAdded {
+			nullViolationAdded = true
+			potentialErrors = append(potentialErrors,
```
should we make a method to add a certain error? this pattern is heavy

ajwerner reviewed (commented):
This code is super non-trivial. Kudos for getting it to work!

Do you have a sense for how often we run into errors? I wonder if it'd be valuable to have the workload output the counters for each type of operation we complete and fail.

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78259)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/workload/schemachange/error_screening.go, line 380 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/78259#-MymGuEr82xuZ8M-MJ7a-r3-380:-N0D62r--3WLQgBkY9Jc:b-6wscv4) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc44a4a5bdd46f3ba58c52820b04a0bb30d1a162/pkg/workload/schemachange/error_screening.go#L380)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I guess we don't add check constraints
</blockquote></details>

If any of the values in a unique constraint have a NULL, then that constraint does not apply to a given row. i.e. The docs say:

You can insert NULL values into columns with the UNIQUE constraint because NULL is the absence of a value, so it is never equal to other NULL values and not considered a duplicate value. This means that it's possible to insert rows that appear to be duplicates if one of the values is NULL.
___
*[pkg/workload/schemachange/error_screening.go, line 397 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/78259#-MymGuEr82xuZ8M-MJ7a-r3-397:-N0Ch6YEBCbLgx0IWa9W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc44a4a5bdd46f3ba58c52820b04a0bb30d1a162/pkg/workload/schemachange/error_screening.go#L397)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

`s/know/known/`? 
</blockquote></details>

Done.
___
*[pkg/workload/schemachange/error_screening.go, line 484 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/78259#-MymGuEr82xuZ8M-MJ7a-r3-484:-N0ChP6s9C-TijON4Wlz:b-wjlx7x) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc44a4a5bdd46f3ba58c52820b04a0bb30d1a162/pkg/workload/schemachange/error_screening.go#L484)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

genearting

Maybe say a tad more about why these might happen?
</blockquote></details>

Reworded this:

// isValidGenerationError these codes can be observed when evaluating values
// for generated expressions. These are errors are not ignored, but added into
// the expected set of errors.
___
*[pkg/workload/schemachange/error_screening.go, line 510 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/78259#-MymGuEr82xuZ8M-MJ7a-r3-510:-N0D5psS3dUvKCWone5G:b91fjz) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc44a4a5bdd46f3ba58c52820b04a0bb30d1a162/pkg/workload/schemachange/error_screening.go#L510)):*
> is this going to cause problems with deadlocks if we've already inserted a row in the same transaction? Should we instead use nested transactions with savepoints and roll back to savepoint?
This is a pseudo transaction implemented via save points. Though as you pointed out schema changes + saved points are not supported. So, let me investigate this, since we didn't trigger that behaviour in long run on roachprod.
___
*[pkg/workload/schemachange/error_screening.go, line 576 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/78259#-MymGuEr82xuZ8M-MJ7a-r3-576:-N0Ci22l3ZYDaf_wmlGv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc44a4a5bdd46f3ba58c52820b04a0bb30d1a162/pkg/workload/schemachange/error_screening.go#L576)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

should we make a method to add a certain error? this pattern is heavy
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
Let me open an issue and start tracking that information. Basically, track the type of operations and error frequencies and dump them out at the end.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78259#-:-N0Dfw9E9XOPB1wAVqie:b-7ps30j)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78259)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/workload/schemachange/error_screening.go, line 510 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/78259#-MymGuEr82xuZ8M-MJ7a-r3-510:-N0Dg5Ws35tHRVkjWc6h:bjxbb0) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc44a4a5bdd46f3ba58c52820b04a0bb30d1a162/pkg/workload/schemachange/error_screening.go#L510)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

> is this going to cause problems with deadlocks if we've already inserted a row in the same transaction? Should we instead use nested transactions with savepoints and roll back to savepoint?
This is a pseudo transaction implemented via save points. Though as you pointed out schema changes + saved points are not supported. So, let me investigate this, since we didn't trigger that behaviour in long run on roachprod.
</blockquote></details>

This is a pseudo transaction implemented via save points. Though as you pointed out schema changes + saved points are not supported. Our save point use is okay since we only do DML inside them, and not execute and DDL. The outer transaction has all of the DDL making it safe to rollback.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> ship it on green. Maybe this deserved a closer read, but I'd rather we merge it. 

> Reviewed 1 of 3 files at r1, 1 of 1 files at r2, 2 of 2 files at r4, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78259#-:-N0m8NFa5po1CBfUzlbd:b-ghkqw3)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/workload/schemachange/error_screening.go` line 508 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/78259#-N0m82Ni98HcKlKAAUjc:-N0m82Ni98HcKlKAAUjd:btvywl3) ([raw file](https://github.com/cockroachdb/cockroach/blob/97fe100e0fc426f7ef56f296dabeda69f9a4f5a8/pkg/workload/schemachange/error_screening.go#L508)):*
> ```Go
> ) (bool, codesWithConditions, error) {
> 	var potentialErrors codesWithConditions
> 	appendPotentialError := func(code pgcode.Code, condition bool) {
> ```

you're always using true here. Should this take a boolean at all?
___
*[`pkg/workload/schemachange/operation_generator.go` line 249 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/78259#-N0m8A4X7yaeHgQ-RNXX:-N0m8A4X7yaeHgQ-RNXY:b-4yc7al) ([raw file](https://github.com/cockroachdb/cockroach/blob/97fe100e0fc426f7ef56f296dabeda69f9a4f5a8/pkg/workload/schemachange/operation_generator.go#L249)):*
> ```Go
> 	setColumnType:           0, // Disabled and tracked with #66662.
> 	survive:                 1,
> 	insertRow:               1,
> ```

what if this was 10?


<!-- Sent from Reviewable.io -->

fqazi committed:
workload/schemachanger: detect if tables already have partitioning

Fixes: #80699

Previously, the schema changer would attempt to convert
a database to multi-region without confirm if any of the
tables were already partitioned. This could lead to
"cannot convert database .. to multiregion" errors.
To address this, this patch add supports for detecting
when the conversion will fail.

Release note: None

fqazi commented:
bors r+

fqazi commented:
bors r+
